version: '3.8'

services:

  db:
    image: postgres:15.1-alpine
    restart: on-failure
    environment:
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      POSTGRES_USER: ${DB_USER}
      POSTGRES_DB: ${DB_NAME}
    volumes:
      - ./postgres_data:/var/lib/postgresql/data
    ports:
      - '${DB_PORT:-5432}:5432'
    healthcheck:
      test: pg_isready -U ${DB_USER} -d ${DB_NAME}
      interval: 3s
      timeout: 3s
      retries: 3

  migrations:
    image: ${IMAGE_TAG}:${GITHUB_REF_NAME}-${GITHUB_RUN_ID}
    depends_on:
      - db
    volumes:
      - ./.env:/todolist/.env
    command: python manage.py migrate

  api:
    image: ${IMAGE_TAG}:${GITHUB_REF_NAME}-${GITHUB_RUN_ID}
    restart: on-failure
    ports:
      - '${API_PORT:-8000}:8000'
    volumes:
      - ./.env:/todolist/.env
    depends_on:
      db:
        condition: service_healthy
      migrations:
        condition: service_completed_successfully
  bot:
    image: ${IMAGE_TAG}:${GITHUB_REF_NAME}-${GITHUB_RUN_ID}
    depends_on:
      db:
        condition: service_healthy
      migrations:
        condition: service_completed_successfully
    volumes:
      - ./.env:/todolist/.env
    command: python manage.py runbot

  frontend:
    image: ${IMAGE_FRONT}
    restart: on-failure
    ports:
      - '${FRONTEND_PORT:-80}:80'
    depends_on:
      - api
